src/libraries/System.Runtime/ref/System.Runtime.cs

```diff
@@ -9113,6 +9113,32 @@ public abstract partial class Module : System.Reflection.ICustomAttributeProvide
         public virtual System.Type ResolveType(int metadataToken, System.Type[]? genericTypeArguments, System.Type[]? genericMethodArguments) { throw null; }
         public override string ToString() { throw null; }
     }
+    public sealed class NullabilityInfoContext
+    {
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("Nullability attributes are trimmed by the linker")]
```

Comment:
1. We shouldn't use the term `linker` in these customer facing messages. `linker` => `trimmer`.
2. Do we have it documented on what the expected behavior is for a trimmed app that is calling these methods? Is there a way that the tooling can see these methods being used and not trim the nullable attributes?

cc @vitek-karas @marek-safar @MichalStrehovsky @agocke

Reply:
One thought I had here is if we could annotate these new `NullabilityInfoContext` APIs with something like:

```C#
[RequiresFeature("System.Reflection.NullabilityInfo")]
public NullabilityInfo Create(ParameterInfo parameterInfo) {}
```

And what this would mean is:

When the linker sees a method with `RequiresFeature` attribute being called, it would "turn on that feature switch".

Then in `ILLink.LinkAttributes.xml`, we would put all these `RemoveAttributeInstances` requests behind the `System.Reflection.NullabilityInfo` feature switch:

https://github.com/dotnet/runtime/blob/eeadfdb2f57e607241b9b09ada1ef08717e618f6/src/mono/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.xml#L133-L165

This would mean, if the feature switch wasn't "on", the attributes could be trimmed safely. If the feature switch was "on", then the attributes won't be trimmed. And then the APIs would work just fine in a trimmed app.

And in the off case where someone was reading the attributes directly, not through these APIs, they could turn the feature switch on for the app in the usual way, and the attributes wouldn't be trimmed.

Thoughts?

Reply:
Another way could be to add `NullabilityInfoContext::IsSupported` and condition the removal of the attributes under that. We would still remove them but there will be standard way to turn it off.

Reply:
A "normal" feature switch would work, but I don't think it gives a great user experience, especially if the default is to remove the attributes.

If a library (say EntityFramework) needs to call these APIs, the consuming app needs to know to turn the feature switch so the attributes are preserved, or else their app is going to behave differently based on whether it is trimmed or not.

Giving a warning to the library isn't really going to help here. What is the library author supposed to do? If they require this information, how else are they supposed to get it?

Reply:
> Giving a warning to the library isn't really going to help here. What is the library author supposed to do? If they require this information, how else are they supposed to get it?

The library author would propagate the information so if the final app requires the API the developer can decide how to handle this. I think the real problem here is how we can scope this because it's very unlikely the app will need BCL nullable annotations.

Reply:
> The library author would propagate the information 

How? With a `RequiresUnreferenceCode` attribute on their API?

> it's very unlikely the app will need BCL nullable annotations.

Agreed. But we currently don't just strip the annotations from the BCL - we strip them from any "IsTrimmable" assembly.

Reply:
This API is for people who are building very advanced runtime generated serializers. They are very likely going to have number of other `RequiresUnreferenceCode` patterns in their code, so adding one more to the list is no big deal.

If you want to use trimming with confidence, you better to not use this API directly nor APIs that use this API indirectly.

Reply:
My rough takeaway here is: this seems like a new scenario for trimming, namely conditionally trimming attributes depending on some particular reflection APIs. At the moment I don't have a good sense of the exact details, but I would like if we could tie the usage of this API to the presence of the nullable attributes.

In other words, the linker currently understands that if you do `GetMethods` on a `typeof`, that's safe, it just means that all the methods on the `Type` are preserved. I would hope we could have something analogous here: using the nullability reflection APIs would mean that the reflection attributes need to be kept on the appropriate members (or all members).

Reply:
The stripping of custom attributes that **we decide** on behalf of the customer as "unimportant" is an illegal optimization that has the potential to break apps _without warnings_. It goes against everything that we were trying to build in .NET 5 and 6. I don't particularly like it.

This is not limited to nullable attributes accessed with this API.

I think we should explore https://github.com/mono/linker/issues/2078. It will be a size regression, but we'll be trading size for correctness and I believe that is a tradeoff we've always been okay with.

Reply:
> (In fact, in talking to @sbomer, this could maybe just be an extra bit in DynamicallyAccessedMemberTypes)

That would be quite a bit of analysis work. We don't track individual members in the analyzer right now. Most of reflection APIs return the members as an IEnumerable or arrays. We would also need to be able to model foreach loops, interface calls, and such. It's doable (.NET Native analyzer can do it) but it adds quite a bit of complexity.

Reply:
For .NET 6 I would very much prefer the feature switch solution. It makes the behavior at least predictable. Currently we remove nullability attributes always (on Mono, CoreCLR doesn't remove them at all) - and the only way to NOT do that is to use descriptors to override this. Having the feature switch would make it more explicit and easier to do.

With the recent changes to handling of IL2026 we could also implement it such that if the feature switch is disabled (nullable attributes are removed) CoreLib produces a single IL2026 warning clearly describing that something in the app needs nullability information at runtime and how to turn it on. It would be similar to what happens with startup hook for example - if the feature switch is on it produces a nice warning.

I do agree with Michal that "hand picked list" of attributes to remove is very problematic. Fortunately this currently only applies to Debugging, HotReload and EventSource attributes which are all behind feature switches - in CoreCLR. In Mono the list is much longer and without any feature switches unfortunately.

Reply:
> I do agree with Michal that "hand picked list" of attributes to remove is very problematic. Fortunately this currently only applies to Debugging, HotReload and EventSource attributes which are all behind feature switches - in CoreCLR. In Mono the list is much longer and without any feature switches unfortunately.

The attribute stripping is going to be aligned between Mono and CoreCLR with https://github.com/dotnet/runtime/issues/48217 that is scheduled to be done in .NET 6 so that people get broken by it equally everywhere.

Note that people do check for nullable attributes manually too because .NET 5 has no such APIs but people have needs to inspect it.

Reply:
@eerhardt 
> 1. We shouldn't use the term linker in these customer facing messages. linker => trimmer.

JFYI i am using a different method instead of `Module.ResolveMethod(MetadataToken)`, therefore removed this attribute

> Do we have it documented on what the expected behavior is for a trimmed app that is calling these methods? 

The expected nullability state for the trimmed app will be `Unknown` as there will be no nullability attribute and no nullability context attribute present, i will add that in the doc

> Is there a way that the tooling can see these methods being used and not trim the nullable attributes?

Looks like that will be covered by  #48217, i think we don't need to add any attribute for this APIs for now

Reply:
> The expected nullability state for the trimmed app will be Unknown as there will be no nullability attribute and no nullability context attribute present, i will add that in the doc

This breaks what we promise with trimming "If there are no warnings, the app will not change behavior after trimming". The above discussed feature switch solution would enable us to attribute the code here in such a way that the developer gets a warning if the nullability attributes were removed and this API was in use.

Reply:
> This breaks what we promise with trimming "If there are no warnings, the app will not change behavior after trimming". The above discussed feature switch solution would enable us to attribute the code here in such a way that the developer gets a warning if the nullability attributes were removed and this API was in use.

Yes, understand, I am not against the feature switch solution, but that can work only for the code user has control over. I meant `Undefined` is expected if the user tries to retrieve nullability info from already trimmed libraries where the user has no control

Reply:
> already trimmed libraries

There's no such thing (well, dotnet/runtime trims the libraries we ship, but that trimming doesn't remove any attributes, and it's very special, nobody else does this) - the only trimming which happens is trimming the entire application. Trimming is always a global operation on the final application - so always in user's control.

Reply:
I guess I should have said for whatever reason if there are no nullability attributes available the resulting NullablityState will be `Unknown`

So @vitek-karas @eerhardt do you prefer to add an attribute to warn for the API usage until the above-discussed feature is resolved? If so what attribute to use? I don't see any reason to hold the PR for that feature implementation

Reply:
As confirmed offline added RequiresUnreferencedCode attribute to the APIs to warn about the trimmer until the safe trimming issue is solved